File: 栈（辅助栈 解最小栈）.py
# 要在 o1 的时间复杂度弹出最小栈，不能用遍历法，核心思路是 用第二个栈 辅助栈：
# 主栈正常存数据，辅助栈 同步记录每一步的最小值
# push时候 辅助栈压入 当前值和栈顶的较小者。getMin时 直接读辅助栈的栈顶。
# 辅助栈 的每一层，都记住了 当时 的最小值，这样确保了不会混乱。

class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        cur_min = val
        if self.min_stack:
            cur_min = min(val, self.min_stack[-1])
        self.min_stack.append(cur_min)

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

File: 栈（辅助栈 解最小栈）.rs
// rust重写 （rust兼具安全性和极致性能）

struct MinStack {
    stack: Vec<i32>,
    min_stack: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack {
            stack: Vec::new(),
            min_stack: Vec::new(),
        }
    }
    
    fn push(&mut self, val: i32) {
        self.stack.push(val);
        
        //使用 match 或 if let 处理空栈
        match self.min_stack.last() {
            Some(&min) => self.min_stack.push(val.min(min)),
            None => self.min_stack.push(val),
        }
    }
    
    fn pop(&mut self) {
        self.stack.pop();
        self.min_stack.pop();
    }
    
    fn top(&self) -> i32 {
        self.stack.last().copied().unwrap()
    }

    fn get_min(&self) -> i32 {
        self.min_stack.last().copied().unwrap()
    }
}
